src/NUnitFramework/framework/Internal/AsyncEnumerableAdapter.cs

```diff
@@ -0,0 +1,29 @@
+// Copyright (c) Charlie Poole, Rob Prouse and Contributors. MIT License - see LICENSE.txt
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+
+namespace NUnit.Framework.Internal
+{
+    internal static partial class AsyncEnumerableAdapter
+    {
+        public static IEnumerable? CoalesceToEnumerable(object? enumerable)
+        {
+            if (enumerable is null)
+            {
+                return null;
+            }
+            else if (TryGetAsyncBlockingEnumerable(enumerable, out var result))
```

Comment:
As `IEnumerable` is more likely, I would check that first. Especially as the .NET Framework version of this method uses reflection.

The code now relies on a runtime InvalidCastException if the method at hand doesn't return an `IEnumerable` at all.
Maybe you should make that more explicit?

Reply:
> As IEnumerable is more likely, I would check that first.

Thanks, will do. The current ordering was a conscious decision, though perhaps misguided, based on the discussion in the AsyncDispose PR. I think it's unlikely a type will implement both interfaces, in particular based on that I can't even find (m)any concrete implementations in the BCL - they're predominantly compiler-generated via `yield return` in an async method.

Reply:
> The code now relies on a runtime InvalidCastException if the method at hand doesn't return an IEnumerable at all.
Maybe you should make that more explicit?

I actually meant to do this but had forgotten! Thanks